specs/eventing/data-plane.md

```diff
@@ -1,125 +1,171 @@
+SHOULD be performed using the CloudEvents HTTP Binding, version 1.0.
+
+Senders MAY probe the recipient with an
+[HTTP OPTIONS request](https://tools.ietf.org/html/rfc7231#section-4.3.7); if
+implemented, the recipient MUST indicate support for the POST verb using the
+[`Allow` header](https://tools.ietf.org/html/rfc7231#section-7.4.1). Senders
+which receive an error when probing with HTTP OPTIONS SHOULD proceed using the
+HTTP POST mechanism.
+
+### Event Acknowledgement and Repeat Delivery
+
+Event recipients MUST use the HTTP response code to indicate acceptance of an
+event. The recipient SHOULD NOT return a response accepting the event until it
+has handled event (processed the event or stored it in stable storage). The
+following response codes are explicitly defined; event recipients MAY also
+respond with other response codes. A response code not in this table SHOULD be
+treated as a retriable error.
+
+| Response code | Meaning                     | Retry | Delivery completed | Error |
+| ------------- | --------------------------- | ----- | ------------------ | ----- |
+| `1xx`         | (Unspecified)               | Yes\* | No\*               | Yes\* |
+| `200`         | [Event reply](#event-reply) | No    | Yes                | No    |
+| `202`         | Event accepted              | No    | Yes                | No    |
+| other `2xx`   | (Unspecified)               | Yes\* | No\*               | Yes\* |
+| other `3xx`   | (Unspecified)               | Yes\* | No\*               | Yes\* |
+| `400`         | Unparsable event            | No    | No                 | Yes   |
+| `404`         | Endpoint does not exist     | Yes   | No                 | Yes   |
+| other `4xx`   | Error                       | Yes   | No                 | Yes   |
+| other `5xx`   | Error                       | Yes   | No                 | Yes   |
+
+\* Unspecified `1xx`, `2xx`, and `3xx` response codes are **reserved for future
+extension**. Event recipients SHOULD NOT send these response codes in this spec
+version, but event senders MUST handle these response codes as errors and
+implement appropriate failure behavior.
+
+<!-- TODO: Should 3xx redirects and 401 (Unauthorized) or 403 (Forbidden) errors
+be retried? What about `405` (Method Not Allowed), 413 (Payload Too Large), 414
+(URI Too Long), 426 (Upgrade Required), 431 (Header Fields Too Large), 451
+(Unavailable for Legal Reasons)? -->
+
+Recipients MUST be able to handle duplicate delivery of events and MUST accept
+delivery of duplicate events, as the event acknowledgement could have been lost
+in return to the sender. Event recipients MUST use the
+[`source` and `id` attributes](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#required-attributes)
+to determine duplicate events (see [observability](#observability) for an
+example case where other event attributes may vary from one delivery attempt to
+another).
```

Comment:
What's the point of specifying here that i need to determine duplicate events if then the action for them is basically "no-op"?

Reply:
> The point is that someone writing an event sink needs to send an acknowledgement for the duplicate event, and can't send (for example) 412 Precondition Failed to indicate to the upstream that they already had that event.

If upstream is a knative component, what should we do with 412? Just don't retry sending the event I guess? And how is that different, from knative perspective, to just any other non-retryable error code?

> If the event sink doesn't want to process duplicate events, they need to have some sort of duplicate detection and tracking mechanism on their own side for doing that

If the event sink is a knative component, how that should work? What you're describing here is a deduplication system that is implicitly requiring some sort of state to persist the `source` and `id` already seen. We haven't tackled at all this discussion in eventing, and it's quite wide because it requires state, some window based system, etc. I think you should remove all this section about duplicated events, because for knative eventing, we just don't have neither specified or implemented any special treatment for them.

Reply:
I too had similar questions in the earlier version of this pr. From that discussion I left with the intention being that _IF_ the sink wants to perform deduping, it must use those fields and can not for example use something else alone. So, if we agree that is the intention of this, then maybe work on wording to make that clearer. If we don't agree on intention then let's clear that up. I do think it's important for the reader of the spec (after reading the "at least once" section) to have an understanding would it be possible (and if so, how) to build a system that could remove duplicates, and I think without something like this (wording changes, etc.) it's not complete.

Reply:
> From that discussion I left with the intention being that IF the sink wants to perform deduping, it must use those fields and can not for example use something else alone.
> I do think it's important for the reader of the spec (after reading the "at least once" section) to have an understanding would it be possible (and if so, how) to build a system that could remove duplicates

TBH all of that is arguable, a proper dedup system IMO should have a certain level of configurability that allows me to decide how to dedup, picking certain fields or even writing an expression for it. You also usually need to bring in the "time variable" to define dedup windows, which is not specified here, nor in eventing we have any concept of "time", if you get what i mean.

Because we haven't anything that allows such configurations in our api, neither we have any implementation that has such dedup mechanism, I prefer to rather avoid having this in the spec because it's a rathole. Let's not specify things we don't implement, neither we have ever really investigated.

Reply:
Similar to this comment https://github.com/knative/specs/pull/24#discussion_r633272358 and this comment https://github.com/knative/specs/pull/24#discussion_r633274832 I see how we could have these topics in a separate document, something like the CE spec _primer_: https://github.com/cloudevents/spec/blob/v1.0.1/primer.md

Reply:
I get what you're saying about the complexities involved and I totes agree. All I think this is saying that you as developer must use source/id because they are supposed to be unique.

The CloudEvents spec says that already:
https://github.com/cloudevents/spec/blob/v1.0.1/primer.md#id

and in particular the last sentence there:

```
...for the purposes of this CloudEvent attribute those meanings are not relevant and therefore using id for some other purpose beyond uniqueness checking is out of scope of the specification and not recommended.
```
says that it is indeed source/id that guarantees uniqueness => you can use that for deduping. So I'm fine dropping the last sentence. 

FYI, here's the earlier discussion:
https://github.com/knative/specs/pull/24#discussion_r629194956

Reply:
What Ville said -- CloudEvents requires that each `source` ensure that `id` is globally-for-all-time unique for events emitted by that source.

If a destination system such as a Knative Service can't handle the events idempotently, then it may need something like a lookaside cache (e.g. Redis) where it can track the events it has already processed. This isn't built into Knative serving, and _is_ a state store that's outside of Knative. That's hardly unique to Knative; AWS Lambda has similar requirements and [solutions](https://medium.com/avmconsulting-blog/deduplicating-amazon-sqs-messages-dc114d1e6545), [2](https://docs.aws.amazon.com/streams/latest/dev/kinesis-record-processor-duplicates.html), [3](https://forums.aws.amazon.com/thread.jspa?threadID=252415).

I'm happy to rewrite this section to make the two points clearer -- would the following help?

> As specified in the [CloudEvents specification](https://github.com/cloudevents/spec/blob/v1.0.1/primer.md#id), event recipients MUST use the [`source` and `id` attributes](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#required-attributes) to determine duplicate events if needed. This specification does not describe state requirements for clients which need to detect duplicate events. (See [observability](#observability) for anexample case where other event attributes may vary from one delivery attempt toanother)

Reply:
I also added a note that idempotency was a reasonable way to deal with duplicate events.

Reply:
This other wording is even worse IMO because it's somehow implying, or at least it's confusing the reader, that event recipients MUST deduplicate events, because you tell the reader that you MUST determine duplicated events. And you don't specify if event recipients is Knative or not.

Given also the cloudevents primer talks about that, I think we should not put ourselves in the business of duplicated events at all and remove all this text block about duplicates. We can always add something later when we'll have more context, or maybe even some facility to deal with them.

Reply:
Just my $.02, to me this is clearer and says "if needed" saying it's not a must.